### ISSUE TITLE
Poll for longRunningRecognize result from another process?

### ISSUE THREAD

**Issue by bkw** [googleapis/nodejs-speech#10]
I am looking for a way to poll for the status of a `longRunningRecognize()` operation from another process.

The Usecase is processing _very_ long audiofiles, when more often than not, the polling within `.promise()` fails and the state (and thus the whole request) is lost. If I had the ability to poll for that status using some serialized state of the original request, I would still be able to retrieve the results.

In other words: I'd like to be able to poll for the status (and retrieve the eventual results) of a long running operation even if the process that started the operation has died. 

Is that possible? Can somebody point me into the right direction?

**Comment by stephenplusplus** [comment]
I'm poking around, and I don't think there's an easy way to do this. I've [created an issue](https://github.com/googleapis/toolkit/issues/1778) on the repo that generates the code for this repo, and it will hopefully expose a method to make this easier.

**Comment by bkw** [comment]
Sorry for the accidental close. Thanks for looking into this!

**Comment by bkw** [comment]
By closing, the tags were removed by wafflebot. Can someone bring them back? Sorry again for the hassle.

**Comment by ismailbaskin** [comment]
This works until the problem is solved.
```js
const { google } = require('googleapis');

(async () => {
  const auth = await google.auth.getClient({
    scopes: ['https://www.googleapis.com/auth/cloud-platform']
  });
  const { data } = await google.speech('v1').operations.get({ auth, name: OPERATION_NAME });

  console.log(data);
})();
```

**Comment by ctataryn** [comment]
FYI, I have a similar use case where in one Google Cloud Function I initiate a `longRunningRecognize`, and I return the `Operation` name from that request.  In another GCF I pass it that `Operation` name, and I poll.  I have the following code in my polling function, however I don't know how to get at the actual underlying data of the response when it comes to the metadata, or the response properties. They are `Buffer`s.

```javascript
exports.checkOperationProgress = (req, res) => {
  const speech = require('@google-cloud/speech');
  const client = new speech.SpeechClient();
  const util = require('util');
  let operationId = req.query.operationId || req.body.operationId;
  if (operationId) {
    return client.operationsClient.getOperation({name: operationId}).then(function(responses) {
	let opResp = responses[0];
        console.log('Buffer: ' + util.inspect(opResp, false, null));
      	res.setHeader('Content-Type', 'application/json');
  		res.status(200).send(JSON.stringify({done: opResp.done, error: opResp.error}));
	}).catch(function(err) {
		console.error(err);
      	res.status(500).send(err);
	});
  } else {
    res.status(404).send();
  }
};
```
To explain further: When the `done` property is true, I'd then like to also return the operation's response property as that'll contain the transcription. I'd also like to pass back the percentage information from the `metadata` property.

In the logs, what I'll see as the output of `console.log('Buffer: ' + util.inspect(response.response.value, false, null));` is something along the lines of:

```javascript
Buffer: { name: '1359558309687340817', metadata: { type_url: 'type.googleapis.com/google.cloud.speech.v1.LongRunningRecognizeMetadata', value: <Buffer 08 64 12 0c 08 c2 d1 f8 e1 05 10 b0 c1 a7 e1 02 1a 0c 08 c2 da f8 e1 05 10 b0 9b b5 ac 01> }, done: true, response: { type_url: 'type.googleapis.com/google.cloud.speech.v1.LongRunningRecognizeResponse', value: <Buffer 12 5e 0a 5c 0a 13 64 65 76 65 6c 6f 70 65 72 20 64 65 76 65 6c 6f 70 65 72 15 4b 59 0a 3f 1a 1f 0a 08 08 06 10 80 ce e4 cd 02 12 08 08 08 10 80 c6 86 ... > }, result: 'response' }
```
But I don't know how to get those buffers into their corresponding types.  Does anyone know how to accomplish this?  Ideally I'd like to dump `opResp.response` out as JSON, but not a JSON representation of the object as a Buffer, I want the actual contents of the underlying object properties if you know what I mean.

**Comment by alexander-fenster** [comment]
Hi @ctataryn,

You'll need to reproduce the magic from

https://github.com/googleapis/gax-nodejs/blob/7535ef2304bbddd4ec4f6cf458bc88556ae30102/src/longrunning.ts#L291-L320

in your code. Both response and metadata are some serialized protobuf types, the exact types are defined in the service configuration (this is how the code generator knows which types to use):

https://github.com/googleapis/googleapis/blob/7a81902f287a4ac2ac130994889c80f87f91d035/google/cloud/speech/v1/speech_gapic.yaml#L85-L87 

and defined here:

https://github.com/googleapis/googleapis/blob/fb44301b2b667c56a0f2a361a5a7d3709d5f6a69/google/cloud/speech/v1/cloud_speech.proto#L369-L388

You can either take the code from `longrunning.ts` from gax package and copy it to your project, or just use [protobufjs](https://www.npmjs.com/package/protobufjs) with the protobuf types from `cloud_speech.proto` to decode your buffers.

**Comment by alexander-fenster** [comment]
@ctataryn Yes. I actually wanted to suggest that as well. This library (`@google-cloud/speech`) is specifically designed to hide all complicated stuff related to long-running operations under the hood, providing a promise that just does everything you need. Since you obviously need to go deeper and expose everything that was hidden on purpose, using a simpler library, `googleapis`, might be a better option. 

One drawback is that `googleapis` use plain old JSON over HTTP/1, compared to protobuf over gRPC (HTTP/2) that this library uses. But if you're not really concerned about performance and JSON parsing overhead, you're good to go with the old package.

**Comment by ismailbaskin** [comment]
@alexander-fenster Actually, a speech operation trigger for GCF would provide much better serverless ML pipeline experience.

**Comment by ctataryn** [comment]
@ismailbaskin that would certainly be ideal for my usage.

**Comment by alexander-fenster** [comment]
AFAIK we cannot set a cloud function trigger for a completion of arbitrary long running operation. I will ask internally if there is such an option and will get back to you folks.

**Comment by sorokinvj** [comment]
Hi guys, would you please help, I guess I have similar issue, but don't know how to handle it yet. 

So, I'm sending a 4 min file to speech via longRunningRecognize, and it breaks it into two operations. The problem is that in my code I resolve results .on("complete) so when second results come back my mongoose driver sends an error

How do I change the code and concat all results from all operations? (I'm building app that hopefully will recognise 1 hour files).

here is my code: 

```
exports.transcribe = async (gcsuri) => {
    process.env.GOOGLE_APPLICATION_CREDENTIALS = process.env.SPEECH_KEY
    const speech = require('@google-cloud/speech').v1p1beta1
    const client = new speech.SpeechClient()
    const audio = {
        uri: gcsuri
    }
    const request = {
        audio: audio,
        config: { encoding: 'OGG_OPUS',
                  sampleRateHertz: 24000,
                  languageCode: 'ru',
                  enableSpeakerDiarization: true,
                  enableWordConfidence: true,
                  maxAlternatives: 1,
                  enableAutomaticPunctuation: true, 
                  enableWordTimeOffsets: true, 
                }
    }

    return await client.longRunningRecognize(request)
        .then(responses => {
                                const [operation, initialApiResponse] = responses

                                return new Promise((resolve, reject) => {

                                    operation.on('progress', (metadata, apiResponse) => {
                                        console.log('metadata', metadata)
                                        console.log("api response", apiResponse)
                                    })

                                    operation.on('error', err => {
                                        throw("long running", err)
                                    })
                                    
                                    operation.on('complete', (result, metadata, finalApiResponse) => {
                                        console.log("complete", finalApiResponse)
                                        resolve({
                                            words: result.results[0].alternatives[0].words,
                                            transcript: result.results[0].alternatives[0].transcript,
                                            confidence: result.results[0].alternatives[0].confidence
                                        })
                                    })
                                })
                })
        .catch(err => {
            console.error("longRunningRecognize", err)
        })
```

**Comment by ctataryn** [comment]
Including the error message would help greatly.

Craig


________________________________

**Comment by sorokinvj** [comment]
@ctataryn Sorry, Craig,  my problem was that I handle results in a wrong way, so I deleted my message to not confuse others. 

Still, my code is somehow spawn two identical operations and I don't know what I did wrong, I guess it is somehow connected to new Promise, but I'm not sure

Here is metadata (when first operation (1143342213005448063) is 79 percent complete here comes another 3846326391202263972) :

```
metadata LongRunningRecognizeMetadata {
   startTime:
    Timestamp {
      seconds: Long { low: 1551571623, high: 0, unsigned: false },
      nanos: 263143000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571623, high: 0, unsigned: false },
      nanos: 765666000 } }
 api response { name: '1143342213005448063',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 12 0b 08 a7 b5 ec e3 05 10 d8 fc bc 7d 1a 0c 08 a7 b5 ec e3 05 10 d0 c5 8c ed 02> },
   done: false }
 metadata LongRunningRecognizeMetadata {
   startTime:
    Timestamp {
      seconds: Long { low: 1551571623, high: 0, unsigned: false },
      nanos: 263143000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571624, high: 0, unsigned: false },
      nanos: 521859000 } }
 api response { name: '1143342213005448063',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 12 0b 08 a7 b5 ec e3 05 10 d8 fc bc 7d 1a 0c 08 a8 b5 ec e3 05 10 b8 df eb f8 01> },
   done: false }
 metadata LongRunningRecognizeMetadata {
   startTime:
    Timestamp {
      seconds: Long { low: 1551571623, high: 0, unsigned: false },
      nanos: 263143000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571627, high: 0, unsigned: false },
      nanos: 554753000 } }
 api response { name: '1143342213005448063',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 12 0b 08 a7 b5 ec e3 05 10 d8 fc bc 7d 1a 0c 08 ab b5 ec e3 05 10 e8 b7 c3 88 02> },
   done: false }
 metadata LongRunningRecognizeMetadata {
   progressPercent: 79,
   startTime:
    Timestamp {
      seconds: Long { low: 1551571623, high: 0, unsigned: false },
      nanos: 263143000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571692, high: 0, unsigned: false },
      nanos: 255757000 } }
 api response { name: '1143342213005448063',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 08 4f 12 0b 08 a7 b5 ec e3 05 10 d8 fc bc 7d 1a 0b 08 ec b5 ec e3 05 10 c8 95 fa 79> },
   done: false }
 
 metadata LongRunningRecognizeMetadata {
   startTime:
    Timestamp {
      seconds: Long { low: 1551571742, high: 0, unsigned: false },
      nanos: 801690000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571743, high: 0, unsigned: false },
      nanos: 603198000 } }
 api response { name: '3846326391202263972',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 12 0c 08 9e b6 ec e3 05 10 90 a3 a3 fe 02 1a 0c 08 9f b6 ec e3 05 10 b0 a4 d0 9f 02> },
   done: false }
 metadata LongRunningRecognizeMetadata {
   startTime:
    Timestamp {
      seconds: Long { low: 1551571742, high: 0, unsigned: false },
      nanos: 801690000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571744, high: 0, unsigned: false },
      nanos: 318200000 } }
 api response { name: '3846326391202263972',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 12 0c 08 9e b6 ec e3 05 10 90 a3 a3 fe 02 1a 0c 08 a0 b6 ec e3 05 10 c0 b1 dd 97 01> },
   done: false }
 metadata LongRunningRecognizeMetadata {
   progressPercent: 79,
   startTime:
    Timestamp {
      seconds: Long { low: 1551571742, high: 0, unsigned: false },
      nanos: 801690000 },
   lastUpdateTime:
    Timestamp {
      seconds: Long { low: 1551571807, high: 0, unsigned: false },
      nanos: 301894000 } }
 api response { name: '3846326391202263972',
   metadata:
    { type_url:
       'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
      value:
       <Buffer 08 4f 12 0c 08 9e b6 ec e3 05 10 90 a3 a3 fe 02 1a 0c 08 df b6 ec e3 05 10 f0 92 fa 8f 01> },
   done: false }
```

**Comment by RobertoC27** [comment]
Is there any progress on this issue?

**Comment by bcoe** [comment]
@ctataryn, @RobertoC27, is your inquiry related to the idea of being able to trigger a cloud function after a long running operation?

Could I request that you open an [issue here](https://issuetracker.google.com/savedsearches/559758), which will in turn be triaged by the product team at Google :smile: 

@bkw with regards to this original issue, did you figure out a solution to your problem eventually? I'm wondering if we can close this old issue.

**Comment by RobertoC27** [comment]
@bcoe yes I ran into the same issue as @ctataryn , creating a longRunningOperation from a cloud function and trying to obtain its result from another cloud function. 

p.d here's the link for issue https://issuetracker.google.com/issues/136100419

**Comment by bcoe** [comment]
@bkw I'm going to go ahead and close this issue, since it's been a couple years and I'm guessing one of the approaches outlined in this ticket might have worked for you :smiley: 

@RobertoC27, thank you for opening that ticket in the issue tracker. That will help make sure that folks on the product team hear about this request.

**Comment by alexander-fenster** [comment]
For everyone who might need it in future, I've just written two solutions for this problem here: https://github.com/googleapis/nodejs-speech/issues/399. I will work on making this go all the way to our official documentation.
